File: tests/parity.js
/* Dump every fixture through the browser worker as JSON, in the same shape as
 * tests/parity.py, so the two grammars can be diffed.
 *
 * Needs the app running (default http://127.0.0.1:8811) and playwright-core.
 * Usage: node tests/parity.js [baseUrl] > out.json
 */
"use strict";

const fs = require("fs");
const path = require("path");

const BASE = process.argv[2] || "http://127.0.0.1:8811";
const PLAYWRIGHT = process.env.PLAYWRIGHT_PATH || "playwright";
const FIXTURES = path.join(__dirname, "fixtures");
const PAGE = 100;

// Minimal page whose only job is to host the worker on the app's origin.
const HARNESS = `<!DOCTYPE html><html><head><meta charset="utf-8"></head><body>
<input type="file" id="f" multiple>
<script>
window.run = function (pageSize) {
  const files = Array.from(document.getElementById('f').files);
  const worker = new Worker('/static/worker.js');
  let seq = 0;
  const waiting = new Map();
  worker.onmessage = e => {
    const m = e.data;
    if (m.t === 'progress' || m.t === 'indexProgress' || m.t === 'searchProgress') return;
    const key = m.t === 'schema' ? 'schema:' + m.id : m.t + ':' + m.id;
    const resolve = waiting.get(key);
    if (resolve) { waiting.delete(key); resolve(m); }
  };
  const send = (msg, expect) => new Promise(res => {
    msg.id = ++seq;
    waiting.set(expect + ':' + msg.id, res);
    worker.postMessage(msg);
  });

  return (async () => {
    const out = [];
    for (const file of files) {
      const opened = await send({ t: 'open', file }, 'schema');
      const tables = [];
      for (const t of opened.schema.tables) {
        const got = await send({ t: 'rows', table: t.name, page: 1, pageSize, query: '' }, 'rows');
        tables.push({
          name: t.name,
          columns: t.columns,
          primary_key: t.primary_key,
          foreign_keys: t.foreign_keys,
          row_count: got.payload.total_rows,
          rows: got.payload.rows,
        });
      }
      out.push({ file: file.name, tables });
    }
    return out;
  })();
};
</script></body></html>`;

(async () => {
  const { chromium } = require(PLAYWRIGHT);
  const browser = await chromium.launch();
  const page = await browser.newPage();
  const errors = [];
  page.on("pageerror", e => errors.push(e.message));

  await page.route(BASE + "/__parity", route =>
    route.fulfill({ status: 200, contentType: "text/html", body: HARNESS }));
  await page.goto(BASE + "/__parity");

  const files = fs.readdirSync(FIXTURES).filter(f => f.endsWith(".sql")).sort()
    .map(f => path.join(FIXTURES, f));
  await page.setInputFiles("#f", files);

  const result = await page.evaluate(size => window.run(size), PAGE);
  await browser.close();

  if (errors.length) {
    console.error("page errors:\n" + errors.join("\n"));
    process.exit(1);
  }
  process.stdout.write(JSON.stringify(result, null, 2));
})().catch(e => { console.error("FAILED", e.message); process.exit(1); });

File: tests/parity.py
"""Dump every fixture through the Python parser as JSON.

The JS worker produces the same shape via tests/parity.js; parity_check.py
compares the two so the grammars cannot drift apart unnoticed.

Usage: python tests/parity.py > out.json
"""

from __future__ import annotations

import json
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
sys.path.insert(0, str(ROOT))

from services import parser  # noqa: E402

FIXTURES = ROOT / "tests" / "fixtures"
PAGE = 100


def describe(path: Path) -> dict:
    with open(path, encoding="utf-8", errors="replace", newline="") as handle:
        dump = parser.parse(handle.read())
    tables = []
    for table in dump.tables:
        columns, rows, total = parser.read_rows(dump, table, 0, PAGE)
        tables.append(
            {
                "name": table.name,
                "columns": [
                    {
                        "name": c.name,
                        "type": c.type,
                        "nullable": c.nullable,
                        "default": c.default,
                        "is_pk": c.is_pk,
                        "is_unique": c.is_unique,
                        "auto_increment": c.auto_increment,
                    }
                    for c in table.columns
                ],
                "primary_key": table.primary_key,
                "foreign_keys": [
                    {"columns": fk.columns, "ref_table": fk.ref_table, "ref_columns": fk.ref_columns}
                    for fk in table.foreign_keys
                ],
                "row_count": total,
                "rows": rows,
            }
        )
    return {"file": path.name, "tables": tables}


def main() -> None:
    sys.stdout.reconfigure(encoding="utf-8")  # fixtures contain non-ASCII rows
    out = [describe(path) for path in sorted(FIXTURES.glob("*.sql"))]
    json.dump(out, sys.stdout, indent=2, ensure_ascii=False, default=str)


if __name__ == "__main__":
    main()
